### PR TITLE
Closes #6530 #6531 #6532 #6534 #6535 #6552: Preload images markup

### LINKED ISSUE #6530: Preload the picture markup
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version => 3.16_alpha1
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
srcset of picture element isn't preloaded as mentioned here https://web.dev/articles/preload-responsive-images#preload_and_picture

**Expected behavior**
srcset shall be preloaded as same as src img when it's detected a LCP image

The markup input:
```
<picture>
    <source srcset="small_cat.jpg" media="(max-width: 400px)">
    <source srcset="medium_cat.jpg" media="(max-width: 800px)">
    <img src="large_cat.jpg">
</picture>
```

The markup output:
```
<link rel="preload" href="small_cat.jpg" as="image" media="(max-width: 400px)" fetchpriority="high">
<link rel="preload" href="medium_cat.jpg" as="image" media="(min-width: 400.1px) and (max-width: 800px)" fetchpriority="high">
<link rel="preload" href="large_cat.jpg" as="image" media="(min-width: 800.1px)" fetchpriority="high">
```

**Acceptance Criteria (for WP Media team use only)**
**LCP**
1. `source srcset` images are preloaded using the image URL from the `srcset`
2.  media attributes must be moved to the preload markup
3.  if the `min-width` is not defined for the next `srcset` images we'll need to add it using the previous `max-width` value + 0.1px.
4.  if the `min-width` is not defined for the `img` markup, we'll need to add it using the previous `max-width`  value + 0.1px.
5. `fetchpriority="high"` is added to the preload markup
6. Image is excluded from LL feature

**Above the fold**
1. Image is excluded from LL Background Images feature

### LINKED ISSUE #6531: Preload the responsive image markup
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
As a part of 3.16 scope we need to preload the responsive image when it's detected as LCP.

https://web.dev/articles/preload-responsive-images#imagesrcset-and-imagesizes

**Expected behavior**
Input markup:
```
 <img src="wolf.jpg" srcset="wolf_400px.jpg 400w, wolf_800px.jpg 800w, wolf_1600px.jpg 1600w" sizes="50vw" alt="A rad wolf">
```

Output markup:
```
<link rel="preload" as="image" href="wolf.jpg" imagesrcset="wolf_400px.jpg 400w, wolf_800px.jpg 800w, wolf_1600px.jpg 1600w" imagesizes="50vw" fetchpriority="high">
```

**Acceptance Criteria (for WP Media team use only)**
**LCP**
1. Image is preloaded using the image URLs from `srcset` attribute and becomes `imagesrcset`
2. `sizes="50vw"` attribute is moved to the preload markup and becomes ` imagesizes="50vw"`
3. `fetchpriority="high"` is added to the preload markup
4. Image is excluded from LL feature

**Above the fold**
1. All Images within responsive markup are excluded from LL feature

### LINKED ISSUE #6532: Preload the single background images markup
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
As a part of 3.16 scope we need to preload the background images when they're detected as LCP.

**Expected behavior**
We'll need to handle 3 different markups, this issue is related to the basic one:

Input:
```
 background-image: url("background.jpg");
```
Output:
```
<link rel="preload" as="image" href="background.jpg" fetchpriority="high">
```

**Acceptance Criteria (for WP Media team use only)**
**LCP**
1. Image is preloaded using the image URL from the declaration
2. `fetchpriority="high"` is added to the preload markup
3. Image is excluded from LL Background Images feature

**Above the fold**
1. Image is excluded from LL Background Images feature

### LINKED ISSUE #6534: Preload the layered background images markup
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
As a part of 3.16 scope we need to preload the layered background images when they're detected as LCP.

**Expected behavior**
Input:
```
background-image: url("img.jpg"), url("img2.jpg");
```
Output:
```
<link rel="preload" as="image" href="img.jpg" fetchpriority="high">
<link rel="preload" as="image" href="img2.jpg" fetchpriority="high">
```

**Acceptance Criteria (for WP Media team use only)**
**LCP**
1. Images are preloaded separately using the image URL from the declaration
2. `fetchpriority="high"` is added to the preload markup
3. Images are excluded from LL Background Images feature

**Above the fold**
1. Images are excluded from LL Background Images feature

### LINKED ISSUE #6535: Preload the image-set background images
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
As a part of 3.16 scope we need to preload the image-set background images when they're detected as LCP.

**Expected behavior**
Input:
```
background-image: image-set( "cat.png" 1x, "cat-2x.png" 2x);
background-image: image-set( url("cat.png") 1x, url("cat-2x.png") 2x);

```
or
```
background-image: -webkit-image-set( "cat.png" 1x, "cat-2x.png" 2x);
background-image: -webkit-image-set( url("cat.png") 1x, url("cat-2x.png") 2x);
```
Output:
```
<link rel="preload" as="image" imagesrcset="cat.png 1x, cat-2x.png 2x" fetchpriority="high">
```

**Acceptance Criteria (for WP Media team use only)**
**LCP**
1. Images are preloaded using the image URL from the `image-set` declaration, they'll be added to the `imagesrcset` attribute
2. `fetchpriority="high"` is added to the preload markup
3. Images are excluded from LL Background Images feature

**Above the fold**
1. Images are excluded from LL Background Images feature

### LINKED ISSUE #6552: Wrong LCP type captured in the DB
**Before submitting an issue please check that you’ve completed the following steps:**
- Using 3.16-pre-alpha
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
When a background image is identified as the LCP, the type in the DB reads `img`

**To Reproduce**
1. Use a background image as the LCP
2. Visit the page
3. See in DB that lcp type is captured as `img`

**Expected behavior**
We should capture background image LCP as `bg-img` instead of `img`

**Screenshots**

**Additional context**

**Acceptance Criteria (for WP Media team use only)**
Clear instructions for developers, to be added before the grooming

### PR DESCRIPTION
**Pull request by Miraeld**
# Description

Fixes #6530 #6531 #6532 #6534 #6535 #6552 

This PR addresses several issues related to image preloading, specifically for images detected as Largest Contentful Paint (LCP) and Above The Fold (ATF) images. The changes ensure that various types of images are correctly preloaded and excluded from lazy loading to improve page load performance. Additionally, it fixes an issue where background images identified as the LCP were incorrectly labeled as 'img' in the database, they are now correctly labeled dynamically.

### User documentation

This update improves the page load performance by correctly preloading various types of images and excluding them from lazy loading when they are detected as Largest Contentful Paint (LCP) and Above The Fold (ATF) images. This means faster loading times for your website visitors.

### Technical documentation

The code works by identifying images that are LCP and ATF, and then preloading these images to improve page load performance. It also correctly labels background images identified as LCP with the corrected type in the database. The changes are implemented across several files, including JavaScript for front-end changes and PHP for back-end changes.

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

## New dependencies
n/a

## Risks
n/a

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Documentation

- [x] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [x] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [x] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style

- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitly.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] I listed the introduced external dependencies explicitly on the PR.
- [x] I validated the repo-specific guidelines from CONTRIBUTING.md.

### CODE REVIEW

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php [hunk 1]
```diff
@@ -91,4 +89,42 @@ public function add_lcp_data() {
...
+	 * @param object $image The image object.
+	 * @param array  $keys  An array of keys in the order of their priority.
+	 *
+	 * @return object|null Returns an object with the 'type' property and the first key that exists in the image object. If none of the keys exist in the image object, it returns null.
+	 */
+	private function create_object( $image, $keys ) {
+		$object       = new \stdClass();
+		$object->type = $image->type;
+
+		if ( 'img-srcset' === $image->type ) {
```
Comment (jeawhanlee) [inline inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: Any reason we don't use a switch statement here?

File: inc/Engine/Media/AboveTheFold/Frontend/Controller.php [hunk 2]
```diff
@@ -271,14 +271,23 @@ private function generate_lcp_link_tag_with_sources( $lcp ): array {
 				}
 				break;
 			case 'picture':
-				if ( ! empty( $lcp->sources ) ) {
-					foreach ( $lcp->sources as $source ) {
-						$sources[] = $source->srcset;
-						$tag      .= $start_tag . 'href="' . $source->srcset . '" media="' . $source->media . '"' . $end_tag;
+				$prev_max_width = null;
+				foreach ( $lcp->sources as $source ) {
+					$media = $source->media;
```
Comment (jeawhanlee) [inline inc/Engine/Media/AboveTheFold/Frontend/Controller.php]: I think we should Isolate these logic in a new method to modularize things for future enhancement for example, when we intend to handle min-width or disordered source tag, this present method can quickly become long and complex.  We can have this logic in a new method that only handles max-width.

**Comment by MathieuLamiot** [comment]
@Miraeld Independently of the review, can you provide some info about test & validation of the content of this PR? I am thinking mostly: what templates has this been tested against to validate the new markups? Is there new tests passing on Rocket-E2e, or tests/templates that should be covered by this that are not passing?
Thanks

**Comment by Miraeld** [comment]
@MathieuLamiot ,
Integrations tests are here to make sure that if we get the right value within the database, the frontend will be correct. 

In terms of detection and saving the data itself (role of the `beacon-lcp.js`), I've only tested manually, as we can't run JS in integrstion tests, sadly. I don't think there are e2e tests even tho some templates are there.

How I've been testing it: 
Simply by using templates available in `rocket-test-data` repo. 

Some are not available yet, so I've built my own on the same structure as we would do for e2e. 
Then do the process manually to make sure everything is detected and saved properly.

**Comment by MathieuLamiot** [comment]
> In terms of detection and saving the data itself (role of the beacon-lcp.js), I've only tested manually, as we can't run JS in integrstion tests, sadly. I don't think there are e2e tests even tho some templates are there.

There should be E2E checking what is saved in DB after a visit on the page, and running on many templates from the QA team. @jeawhanlee should be able to show you, that's what he implemented last cooldown sprint.

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php (hunk 1 above)
Comment (Miraeld) [reply inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: There was a `if` before, so I continued with that logic, 
I will change it in the next commit if it makes it easier ;)

**Comment by Miraeld** [comment]
@MathieuLamiot totally agreed, 
I was planning to look at e2e, as in the same time it's part of the testing we need to do for 3.16. 
It's just I did not want to hold this PR while doing the e2e as e2e is on another repo.

**Review by jeawhanlee — APPROVED** [review]

**Comment by MathieuLamiot** [comment]
As discussed [here](https://github.com/wp-media/wp-rocket/issues/6537#issuecomment-2061387690), this PR is likely to miss the AC about ATF exclusion of Lazyload for elements that are not "img" (so video, picture, p, main, div)

**Comment by Miraeld** [comment]
Well I can add them here. I may be on the path to have a solution but need some optimization :)

**Pull request merged**